command/agent_test.go

```diff
@@ -1691,3 +1695,304 @@ vault {
+	// a successful render unless error_on_missing_key is specified, whereas a
+	// missing secret (4xx response) always results in an error.
+	missingKeyTemplateContent := `{{- with secret "secret/otherapp"}}{"secret": "other",
+{{- if .Data.data.foo}}"foo":"{{ .Data.data.foo}}"{{- end }}}
+{{- end }}`
+	missingKeyTemplateRender := `{"secret": "other",}`
+
+	badTemplateContent := `{{- with secret "secret/non-existent"}}{"secret": "other",
+{{- if .Data.data.foo}}"foo":"{{ .Data.data.foo}}"{{- end }}}
+{{- end }}`
+
+	testCases := map[string]struct {
+		exitOnRetryFailure        *bool
+		templateContents          string
+		expectTemplateRender      string
+		templateErrorOnMissingKey bool
+		expectExit                bool
+		expectError               bool
+	}{
+		"true, no template error": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(true),
+			templateContents:          templateContents(0),
+			expectTemplateRender:      templateRendered(0),
+			templateErrorOnMissingKey: false,
+			expectExit:                true,
+			expectError:               false,
+		},
+		"true, with non-existent secret": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(true),
+			templateContents:          badTemplateContent,
+			expectTemplateRender:      "",
+			templateErrorOnMissingKey: false,
+			expectExit:                true,
+			expectError:               true,
+		},
+		"true, with missing key": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(true),
+			templateContents:          missingKeyTemplateContent,
+			expectTemplateRender:      missingKeyTemplateRender,
+			templateErrorOnMissingKey: false,
+			expectExit:                true,
+			expectError:               false,
+		},
+		"true, with missing key, with error_on_missing_key": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(true),
+			templateContents:          missingKeyTemplateContent,
+			expectTemplateRender:      "",
+			templateErrorOnMissingKey: true,
+			expectExit:                true,
+			expectError:               true,
+		},
+		"false, no template error": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(false),
+			templateContents:          templateContents(0),
+			expectTemplateRender:      templateRendered(0),
+			templateErrorOnMissingKey: false,
+			expectExit:                false,
+			expectError:               false,
+		},
+		"false, with non-existent secret": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(false),
+			templateContents:          badTemplateContent,
+			expectTemplateRender:      "",
+			templateErrorOnMissingKey: false,
+			expectExit:                false,
+			expectError:               true,
+		},
+		"false, with missing key": {
+			exitOnRetryFailure:        pointerutil.BoolPtr(false),
+			templateContents:          missingKeyTemplateContent,
+			expectTemplateRender:      missingKeyTemplateRender,
+			templateErrorOnMissingKey: false,
+			expectExit:                false,
```

Comment:
Same here, this seems like a behavior change on non-error cases

Reply:
Refer to the comment above.